crates/sparrow-runtime/src/stores/object_store_url.rs

```diff
@@ -0,0 +1,271 @@
+use std::{path::PathBuf, str::FromStr};
+
+use error_stack::{IntoReport, ResultExt};
+use serde::{Deserialize, Serialize};
+
+use tokio_util::io::StreamReader;
+use url::Url;
+
+use itertools::Itertools;
+
+use crate::ObjectStoreRegistry;
+
+use super::object_stores::Error;
+
+/// A string referring to a file in an object store.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct ObjectStoreUrl {
+    url: Url,
+}
+
+impl ObjectStoreUrl {
+    pub fn path(&self) -> error_stack::Result<object_store::path::Path, Error> {
+        object_store::path::Path::parse(self.url.path())
+            .into_report()
+            .change_context_lazy(|| Error::UrlInvalidPath(self.url.clone()))
+    }
+
+    pub fn key(&self) -> error_stack::Result<ObjectStoreKey, Error> {
+        match self.url.scheme() {
+            "file" => Ok(ObjectStoreKey::Local),
+            "mem" => Ok(ObjectStoreKey::Memory),
+            // S3 is the traditional S3 prefix for reading from S3.
+            // S3a is the protocol designed for scalability with Hadoop reading in mind.
+            // See: https://aws.amazon.com/blogs/opensource/community-collaboration-the-s3a-story/
+            "s3" | "s3a" => {
+                let bucket = self
+                    .url
+                    .host_str()
+                    .ok_or_else(|| Error::UrlMissingHost(self.url.clone()))?
```

Comment:
Wonder if it's worth introducing `lazy_format!` for cases like these to avoid the clone